packages/twenty-front/src/modules/workflow/workflow-diagram/components/WorkflowDiagramCanvasBase.tsx

```diff
@@ -186,19 +206,214 @@ export const WorkflowDiagramCanvasBase = ({
   );
 
   const { setEdgeHovered, clearEdgeHover } = useEdgeState();
+  const { isInRightDrawer } = useContext(ActionMenuContext);
+  const { rightDrawerState } = useRightDrawerState();
 
   const [workflowDiagramFlowInitialized, setWorkflowDiagramFlowInitialized] =
     useState<boolean>(false);
+  const baseNodes = useMemo(
+    () => (isDefined(workflowDiagram) ? workflowDiagram.nodes : []),
+    [workflowDiagram],
+  );
+  const baseEdges = useMemo(
+    () => (isDefined(workflowDiagram) ? workflowDiagram.edges : []),
+    [workflowDiagram],
+  );
+
+  type PendingPlaceholder = {
+    parentStepId: string;
+    position: { x: number; y: number };
+    sourceHandleId: string;
+    existingTargets: string[];
+  };
 
-  const { nodes, edges } = useMemo(() => {
-    if (isDefined(workflowDiagram)) {
-      return workflowDiagram;
+  const [pendingPlaceholder, setPendingPlaceholder] = useState<
+    PendingPlaceholder | undefined
+  >(undefined);
+
+  const hasActiveInsert =
+    isDefined(workflowInsertStepIds.parentStepId) &&
+    isDefined(workflowInsertStepIds.position);
+
+  useEffect(() => {
```

Comment:
We prefer not to use `useEffect` in the codebase. Most of the time, you can put the logic in an event listener that's called synchronously. Dealing with `useEffects` usually creates bugs. Please try to find the best event listener to call this logic synchronously.